ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimization.java

```diff
@@ -736,13 +729,124 @@ static void applyPreventiveResultsForAutoOrCurativeRangeActions(Network network,
+                    Map<State, Set<Double>> praResults = setPointResults.get(pra);
+                    double craSetPoint = result.getOptimizedSetpoint(cra, state);
+                    praResults.putIfAbsent(state, new HashSet<>(Set.of(craSetPoint)));
+                    praResults.get(state).add(craSetPoint);
+                }
+            }));
+        });
+        multipleInstantRangeActions.forEach(ra -> {
+            setPointResults.put(ra, new HashMap<>(Map.of(preventiveState, Set.of(firstPreventiveResult.getOptimizedSetpoint(ra, preventiveState)))));
+            contingencyResults.forEach((state, result) -> {
+                if (isRangeActionAvailableInState(ra, state, crac)) {
+                    setPointResults.get(ra).put(state, Set.of(result.getOptimizedSetpoint(ra, state)));
+                }
+            });
+        });
+        return setPointResults;
+    }
+
+    /**
+     * Gathers every range action that should not be considered in the second preventive if those 2 criterion are met :
+     * 1- The range action has the same tap in preventive and in a contingency scenario.
+     * 2- For the given state, the crac has limiting RaUsageLimits.
+     */
+    private static Set<RangeAction<?>> getRangeActionsToRemove(Crac crac, State preventiveState, Map<RangeAction<?>, Map<State, Set<Double>>> setPointResults) {
+        Set<RangeAction<?>> rangeActionsToRemove = new HashSet<>();
+        setPointResults.forEach((ra, spMap) -> {
+            double referenceSetPoint = spMap.get(preventiveState).iterator().next();
+            for (Map.Entry<State, Set<Double>> entry : spMap.entrySet()) {
+                State state = entry.getKey();
+                if (!state.isPreventive() && entry.getValue().contains(referenceSetPoint)) {
+                    Instant instant = state.getInstant();
+                    if (crac.getRaUsageLimitsPerInstant().containsKey(instant)) {
+                        String operator = ra.getOperator();
+                        RaUsageLimits raUsageLimits = crac.getRaUsageLimits(instant);
+                        Set<RangeAction<?>> potentiallyAvailableRas = crac.getPotentiallyAvailableRangeActions(state);
+                        long potentiallyAvailableRasForTheOperator = operator == null ? Integer.MAX_VALUE : potentiallyAvailableRas.stream().filter(rangeAction -> {
+                            String otherOperator = rangeAction.getOperator();
+                            return otherOperator != null && otherOperator.equals(operator);
+                        }).count();
+                        int limitingValueForTheTso = Math.min(raUsageLimits.getMaxPstPerTso().getOrDefault(operator, Integer.MAX_VALUE), raUsageLimits.getMaxRaPerTso().getOrDefault(operator, Integer.MAX_VALUE));
+                        if (raUsageLimits.getMaxRa() < potentiallyAvailableRas.size() ||
+                                limitingValueForTheTso < potentiallyAvailableRasForTheOperator) {
```

Comment:
should you also check maxTso limit?

Reply:
Yes I didn't think about that but you're right.

Reply:
I've also realized I should take into account activated network actions in my calculation.